platform/msm8916/msm8916-clock.c

```diff
@@ -684,14 +684,150 @@ static struct clk_lookup msm_clocks_8916[] =
+#define C0_PLL_CONFIG_CTL_CUSEL_250FF			BIT(16)
+#define C0_PLL_CONFIG_CTL_RESERVE_BIT14			BIT(14)
+#define C0_PLL_CONFIG_CTL_CFG_LOCKDET_2PI15		BIT(12)
+#define C0_PLL_CONFIG_CTL_RESERVE_BITS10_0_MAGIC	(BIT(10) | BIT(9) | \
+							 BIT(8)  | BIT(6) | \
+							 BIT(5)  | BIT(2) | BIT(0))
+
+#define C0_PLL_CONFIG_CTL_CFG_RESERVE_BIT30		BIT(30)
+#define C0_PLL_CONFIG_CTL_CFG_RESERVE_BIT14		BIT(14)
+#define C0_PLL_CONFIG_CTL_CFG_LOCKDET_2PI15		BIT(12)
+
+#define APCS_ALIAS0_CMD_RCGR_BASE			0xb111050
+#define APCS_ALIAS0_CMD_RCGR_REG			0x0
+#define APCS_ALIAS0_CMD_RCGR_UPDATE			BIT(0)
+#define APCS_ALIAS0_CFG_REG				0x4
+#define APCS_ALIAS0_CFG_RCGR_SRC_SEL_CLK3		BIT(10)
+#define APCS_ALIAS0_CFG_RCGR_SRC_DIV_2			(BIT(1) | BIT(0))
+#define APCS_ALIAS0_CORE_CBCR_REG			0x8
+#define APCS_ALIAS0_CORE_CBCR_SW_CTL			BIT(0)
+
+static void msm8939_clock_c0_enable_sr2_pll(void)
+{
+	uint32_t base = APCS_C0_PLL_BASE;
+	int regval;
+
+	/* Disable Mode */
+	writel(0, base + C0_PLL_MODE_REG);
+
+	/* Configure L/M/N values - these are accumulator fields */
+	writel(52, base + C0_PLL_L_VAL_REG);
+	writel(0,  base + C0_PLL_M_VAL_REG);
+	writel(1,  base + C0_PLL_N_VAL_REG);
+
+	/* Configure USER_CTL and CONFIG_CTL value */
+	writel(C0_PLL_USER_CTL_MN_EN |
+	       C0_PLL_USER_CTL_PLLOUT_LV_EARLY |
+	       C0_PLL_USER_CTL_PLLOUT_LV_BIST |
+	       C0_PLL_USER_CTL_PLLOUT_LV_AUX |
+	       C0_PLL_USER_CTL_PLLOUT_LV_MAIN,
+	       base + C0_PLL_USER_CTL_REG);
+
+	/*
+	 * The downstream write value is 0x4c015765 which sets several reserved
+	 * bits to unknown values
+	 * 0x4c015765 => 1 0 0 11 00 0 0 00 00 01 0 1 01 0 11101100101
+	 * RESERVE_BIT_30 : 1
+	 * FORCE_PFD_UP |29| 0
+	 * FORCE_PFD_DOWN |28| 0
+	 * NMOSC_FREQ_CTRL |27:26| 11
+	 * PFD_DZSEL |25:24| 00
+	 * NMOSC_EN |23| 0
+	 * RESERVE_BIT22 |22| 0
+	 * ICP_DIV |21:20| 00
+	 * IREG_DIV |19:18| 00
+	 * CUSEL |17:16| 01
+	 * REF_MODE |15| 0
+	 * RESERVE_BIT14 |14| 1
+	 * CFG_LOCKDET |13:12| 01
+	 * FORCE_ISEED |11| 0
+	 * RESERVE_BITS10_0 |0:10| 11101100101
+	 */
+	writel(C0_PLL_CONFIG_CTL_RESERVE_BIT_30	|
+	       C0_PLL_CONFIG_CTL_NMOSC_FREQ_CTRL_PFD_DZSEL_4X |
+	       C0_PLL_CONFIG_CTL_CUSEL_250FF |
+	       C0_PLL_CONFIG_CTL_RESERVE_BIT14 |
+	       C0_PLL_CONFIG_CTL_CFG_LOCKDET_2PI15 |
+	       C0_PLL_CONFIG_CTL_RESERVE_BITS10_0_MAGIC, base + C0_PLL_CONFIG_CTL_REG);
+
+	/* Enable PLL now */
+	writel(C0_PLL_MODE_PLL_BYPASSNL, base + C0_PLL_MODE_REG);
+	udelay(2);
+	writel(C0_PLL_MODE_PLL_RESET_N |
+	       C0_PLL_MODE_PLL_BYPASSNL, base + C0_PLL_MODE_REG);
+	udelay(50);
+	writel(C0_PLL_MODE_PLL_RESET_N |
+	       C0_PLL_MODE_PLL_BYPASSNL |
+	       C0_PLL_MODE_PLL_OUTCTRL,
+	       base + C0_PLL_MODE_REG);
+}
+
+static void msm8939_clock_c0_init(void)
+{
+	uint32_t base;
+	int regval;
+	int count;
+
+	msm8939_clock_c0_enable_sr2_pll();
+
+	base = APCS_ALIAS0_CMD_RCGR_BASE;
+
+	/* Source GPLL0 and 1/2 the rate of GPLL0 0x0403 */
+	regval = APCS_ALIAS0_CFG_RCGR_SRC_SEL_CLK3 |
```

Comment:
Hm, here the clock is set to be sourced from GPLL0 and it gets to `800 * 1/2 = 400 MHz`, I'm unsure if the sr2 pll is needed in this case?

Reply:
Exactly. It might make sense to configure `C0_PLL_USER_CTL_REG` and `C0_PLL_CONFIG_CTL_REG` in lk2nd (just to avoid having to do this upstream). But I don't see why we should set L/M/N or enable the PLL.

I suspect this is just a strange "artifact" from downstream that is actually not necessary.

Reply:
My answer is "that's what downstream does" but 

doing this

`+       //msm8939_clock_c0_enable_sr2_pll();`

still results in this

`a53mux@b111000              1        1        0   400000000          0     0  50000         Y`
`a53mux@b011000              1        1        0   800000000          0     0  50000         Y`

so I guess you are both right that this is dead code for starting the cluster ...

I assume it does something but, I'm happy enough to drop it since its not apparently required for this.